### ISSUE TITLE
how to determine the hash size?

### ISSUE THREAD

**Issue by demeiyan** [facebookresearch/DrQA#180]
Hi Adam,
The paper give hash size 2^24,but if I have 40000 Chinese sentence ,I should how to determine the hash size. The paper [Weinberger et al.,2009](https://arxiv.org/pdf/0902.2206.pdf) is very obscure for me . Could you explain it briefly?

Thanks,
Demei Yan

**Comment by demeiyan** [comment]
40000 Chinese paragraphs

**Comment by ajfisch** [comment]
TL;DR: The hash size is not that sensitive, and you can likely just leave it as is.

The reason we use it is for efficiency (speed and storage). It reduces the dimensionality of the feature space when we use a relatively large vocabulary (unigrams and bigrams).

You might worry that when we use hashing, collisions will hurt. That is, the feature vector is no longer "one hot" in dimension i. It is actually an indicator function over all the possible vocabulary items that hash to i. So a dot product between the vector representations of two documents might be different than the true value.

The Weinberger paper shows that this "distortion error" is low with high probability, and therefore doesn't hurt performance that much (even when the compression ratio/collision rate is high).

**Comment by demeiyan** [comment]
Thanks for your careful explanation !

 I guess the row of matrix is the hash value of the words(unigrams or bigrams),the col of matrix is the doc(doc1 ,doc2,...,docn).So it will be good if the hash size more than the total hash value of the words . Is my opinion correct?

**Comment by demeiyan** [comment]
sorry , I don't attention "even when the compression ratio/collision rate is high" . I know. Thank you !

**Comment by ajfisch** [comment]
Your interpretation of the matrix is correct. But I'm not sure what you mean by the "total hash value of the words."

In general the total number of unique words (union of unigrams and bigrams etc) will be N. So if you didn't hash things, you would enumerate the words 1,...,N (and the matrix would be N x D). Hashing enumerates this modulo M instead (and the matrix would be M x D).

It is ok if M << N.

**Comment by demeiyan** [comment]
I mean the total number of unique words.

**Comment by demeiyan** [comment]
ok，I understand, thanks!
